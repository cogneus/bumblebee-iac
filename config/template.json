{
  "regionCode": "${regionCodes.${region}}",
  "github": {
    "branch": "${branch}"
  },
  "prefix": {
    "ssm": "/${costCenter}/${productName}/${environmentName}",
    "name": "${costCenter}-${productName}-${environmentName}",
    "qual": "${productRef}${environmentName}"
  },
  "s3": {
    "templateBucketSuffix": "template-bucket",
    "artifactBucket": "${prefix.name}-${defaultStage}-${regionCode}-${components.platform}-artifact-bucket",
    "cdkBucket": "${prefix.name}-${defaultStage}-${regionCode}-${components.platform}-cdk-bucket",
    "templateBucket": "${prefix.name}-${stage}-${regionCode}-${components.app}-${s3.templateBucketSuffix}"
  },
  "ssm": {
    "kmsArn": "${prefix.ssm}/${defaultStage}/${components.platform}/cmk/arn",
    "githubToken": "${prefix.ssm}/${defaultStage}/${components.platform}/secrets/github-token",
    "replicationRole": "${prefix.ssm}/${defaultStage}/${components.app}/roles/s3cr",
    "activeStage": "${prefix.ssm}/${defaultStage}/active-stage"
  },
  "secrets": {
    "prefix": "arn:aws:secretsmanager:${region}:${awsAccountId}:secret:${prefix.name}"
  },
  "tags": {
    "costCenter": "${costCenter}",
    "envName": "${environmentName}",
    "productName": "${productName}",
    "componentName": "${componentName}",
    "owner": "${ownerEmail}"
  },
  "iam": {
    "roleArnPrefix": "arn:aws:iam::${awsAccountId}:role",
    "s3crRole": "${prefix.name}-${stage}-${componentName}-s3cr-role",
    "permissionsBoundary": "arn:aws:iam::${awsAccountId}:policy/${permissionsBoundaryName}",
    "codebuildRole": "${prefix.name}-${defaultStage}-${components.platform}-codebuild-role",
    "codepipelineRole": "${prefix.name}-${defaultStage}-${components.platform}-codepipeline-role",
    "cloudformationRole": "${prefix.name}-${defaultStage}-${components.platform}-cloudformation-role",
    "kmsRole": "${prefix.name}-${defaultStage}-${components.platform}-kms-role",
    "githubConnectionRole": "${prefix.name}-${defaultStage}-${components.platform}-github-connection-role",
    "iamRootArn": "arn:aws:iam::${awsAccountId}:root",
    "adminRole": "${sso.adminRole}"
  }
}
