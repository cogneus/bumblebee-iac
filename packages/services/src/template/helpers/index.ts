export * from './keys.helper'
export * from './validate-template.helper'
export * from './generate-id'
export * from './filter-input.class'
