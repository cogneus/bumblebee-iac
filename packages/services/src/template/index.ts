export * from './add'
export * from './get'
export * from './query'
export * from './remove'
export * from './update'
export * from './versions'
