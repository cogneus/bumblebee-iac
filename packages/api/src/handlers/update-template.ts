import { apiGatewayWrapper, warmupWrapper, standardEnvVars, Token } from 'bumblebee-common'
import { TemplateItem, TemplateVersion, updateTemplate } from 'bumblebee-services'
import { checkAPIEnvironmentVars, checkWritePermissions, getClientOptions } from '../helpers'
import { TemplatesParams } from '../types'

checkAPIEnvironmentVars()

export const main = warmupWrapper(
  apiGatewayWrapper<TemplateItem, TemplatesParams, Token, TemplateVersion>({
    handler: ({ payload, params: { id }, serviceContext }) =>
      updateTemplate(serviceContext, id!, payload),
    ...standardEnvVars(),
    serviceName: 'template-api',
    logName: 'update-template',
    getClientOptions,
    authCheck: checkWritePermissions,
  })
)
