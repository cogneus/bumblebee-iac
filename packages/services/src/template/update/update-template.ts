import { ServiceContext, ensureClient, IDBClient } from 'bumblebee-common'
import { TemplateItem, TemplateVersion } from '../../types'
import { validateTemplate } from '../helpers'
import { getTemplate } from '../get/get-template'
import { createCommands } from './create-commands'

export const updateTemplate = async (
  context: ServiceContext,
  id: string,
  template: TemplateItem
): Promise<TemplateVersion> => {
  validateTemplate(template)
  const currentTemplate = await getTemplate(context, id)
  const { version, created } = currentTemplate
  const db = ensureClient<IDBClient>(context, 'db')
  const modified = new Date().toISOString()
  const newVersion: TemplateVersion = {
    ...template,
    id,
    version: version + 1,
    modified,
    created,
  }
  const params = createCommands(newVersion, db.tables.templates.name)
  await db.transactWrite(params)
  return newVersion
}
