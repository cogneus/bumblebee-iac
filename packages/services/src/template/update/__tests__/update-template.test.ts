import { serviceContext } from '../../__mocks__/service-context.mock'
import { dbTemplateItem, templateItem } from '../../__mocks__/template.mock'
import { updateTemplate } from '../update-template'

describe('Update template tests', () => {
  beforeEach(() => {
    jest.clearAllMocks()
    jest.useFakeTimers()
    jest.setSystemTime(new Date('2023-01-01T00:00:00Z'))
  })

  test('db request should match snapshot', async () => {
    let input = null
    ;(serviceContext.clients.db?.get as jest.Mock).mockImplementationOnce(() => dbTemplateItem)
    ;(serviceContext.clients.db?.transactWrite as jest.Mock).mockImplementationOnce((params) => {
      input = params
      return null
    })
    const result = await updateTemplate(serviceContext, dbTemplateItem.id, templateItem)
    expect({
      input,
      result,
    }).toMatchSnapshot()
  })
})
