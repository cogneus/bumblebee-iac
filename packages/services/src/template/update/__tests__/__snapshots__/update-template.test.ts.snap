// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Update template tests db request should match snapshot 1`] = `
{
  "input": [
    {
      "Update": {
        "ExpressionAttributeNames": {
          "#PKGSI": "PKGSI",
          "#SKGSI": "SKGSI",
          "#created": "created",
          "#group": "group",
          "#id": "id",
          "#modified": "modified",
          "#source": "source",
          "#target": "target",
          "#template": "template",
          "#version": "version",
        },
        "ExpressionAttributeValues": {
          ":PKGSI": "group#mock-group",
          ":SKGSI": "modified#2023-01-01T00:00:00.000Z",
          ":created": "2023-01-01T00:00:00Z",
          ":group": "mock-group",
          ":id": "00000000-0000-0000-0000-000000000000",
          ":modified": "2023-01-01T00:00:00.000Z",
          ":source": "mock-source",
          ":target": "mock-target",
          ":template": {},
          ":version": 2,
        },
        "Key": {
          "PK": "template#00000000-0000-0000-0000-000000000000",
          "SK": "current",
        },
        "TableName": "name-templates",
        "UpdateExpression": "SET #group = :group, #source = :source, #target = :target, #template = :template, #id = :id, #version = :version, #modified = :modified, #created = :created, #PKGSI = :PKGSI, #SKGSI = :SKGSI",
      },
    },
    {
      "Put": {
        "ConditionExpression": "attribute_not_exists(#pk) and attribute_not_exists(#sk)",
        "ExpressionAttributeNames": {
          "#pk": "PK",
          "#sk": "SK",
        },
        "Item": {
          "PK": "template#00000000-0000-0000-0000-000000000000",
          "PKGSI": "template#00000000-0000-0000-0000-000000000000",
          "SK": "version#0000000002",
          "SKGSI": "modified#2023-01-01T00:00:00.000Z",
          "created": "2023-01-01T00:00:00Z",
          "group": "mock-group",
          "id": "00000000-0000-0000-0000-000000000000",
          "modified": "2023-01-01T00:00:00.000Z",
          "source": "mock-source",
          "target": "mock-target",
          "template": {},
          "version": 2,
        },
        "TableName": "name-templates",
      },
    },
  ],
  "result": {
    "created": "2023-01-01T00:00:00Z",
    "group": "mock-group",
    "id": "00000000-0000-0000-0000-000000000000",
    "modified": "2023-01-01T00:00:00.000Z",
    "source": "mock-source",
    "target": "mock-target",
    "template": {},
    "version": 2,
  },
}
`;
